File: langs/clojure/tetris-polylith/components/move/src/tetrisanalyzer/move/move.clj
(ns tetrisanalyzer.move.move
  (:require [tetrisanalyzer.piece.interface :as piece]))

(defn cell [board x y [cx cy]]
  (or (get-in board [(+ y cy) (+ x cx)])
      piece/X))

(defn valid-move? [board x y p rotation shapes]
  (every? zero?
          (map #(cell board x y %)
               (piece/piece p rotation shapes))))

(defn left [board x y p rotation _ shapes]
  (when (valid-move? board (dec x) y p rotation shapes)
    [(dec x) y rotation]))

(defn right [board x y p rotation _ shapes]
  (when (valid-move? board (inc x) y p rotation shapes)
    [(inc x) y rotation]))

(defn down
  "Returns [down-move placement] where:
   - down-move: next move when moving down or nil if blocked
   - placement: final placement if blocked, or nil if can move down"
  [board x y p rotation _ shapes]
  (if (valid-move? board x (inc y) p rotation shapes)
    [[x (inc y) rotation] nil]
    [nil [[x y rotation]]]))

(defn rotate [board x y p rotation bitmask shapes]
  (let [new-rotation (bit-and (inc rotation) bitmask)]
    (when (valid-move? board x y p new-rotation shapes)
      [x y new-rotation])))

(defn rotate-with-kick [board x y p rotation bitmask shapes]
  (or (rotate board x y p rotation bitmask shapes)
      (rotate board (dec x) y p rotation bitmask shapes)))

(defn rotation-fn [rotation-kick?]
  (if rotation-kick?
    rotate-with-kick
    rotate))

File: langs/python/tetris-polylith-uv/components/tetrisanalyzer/piece/move/move.py
from tetrisanalyzer.piece import core as piece


def cell(board, x, y, cx, cy):
    yy = y + cy
    xx = x + cx
    try:
        if yy < 0 or xx < 0:
            return piece.X
        return board[yy][xx]
    except IndexError:
        return piece.X


def is_valid_move(board, x, y, p, rotation, shapes):
    return all(cell(board, x, y, *c) == 0 for c in piece.piece(p, rotation, shapes))


def left(board, x, y, p, rotation, _bitmask, shapes):
    return [x - 1, y, rotation] if is_valid_move(board, x - 1, y, p, rotation, shapes) else None


def right(board, x, y, p, rotation, _bitmask, shapes):
    return [x + 1, y, rotation] if is_valid_move(board, x + 1, y, p, rotation, shapes) else None


def down(board, x, y, p, rotation, _bitmask, shapes):
    """Return (down_move, placement).

    - down_move: next move when moving down, or None if blocked
    - placement: [[x, y, rotation]] if blocked, otherwise None
    """

    if is_valid_move(board, x, y + 1, p, rotation, shapes):
        return [x, y + 1, rotation], None
    return None, [[x, y, rotation]]


def rotate(board, x, y, p, rotation, bitmask, shapes):
    new_rotation = (rotation + 1) & bitmask
    return [x, y, new_rotation] if is_valid_move(board, x, y, p, new_rotation, shapes) else None


def rotate_with_kick(board, x, y, p, rotation, bitmask, shapes):
    return (
        rotate(board, x, y, p, rotation, bitmask, shapes) or
        rotate(board, x - 1, y, p, rotation, bitmask, shapes)
    )


def rotation_fn(rotation_kick):
    return rotate_with_kick if rotation_kick else rotate
